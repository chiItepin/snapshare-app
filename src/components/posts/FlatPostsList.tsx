import React, { FunctionComponent } from 'react';
import moment from 'moment';
import {
  View,
  FlatList,
} from 'react-native';
import {
  Card,
  Button,
  Text,
} from 'react-native-ui-lib';
import { NavigationProp } from '@react-navigation/native';
import styles from '../../styles/GlobalStyles';
import HelperStyles from '../../styles/HelperStyles';
import IPost from '../../screens/templates/post';
import RootScreenParams from '../../screens/RootScreenParams';

interface IProps {
  posts: IPost[];
  navigation: any;
  renderHeader: () => React.ReactElement|null;
  renderFooter: () => React.ReactElement|null;
  loaded: boolean;
  handleGetPosts: (loadMore: boolean, page?: number) => void;
}

interface IRenderPostProps {
  item: IPost;
  navigation: NavigationProp<RootScreenParams>;
}

const RenderPost: FunctionComponent<IRenderPostProps> = ({
  item,
  navigation,
}: IRenderPostProps) => (
  <View key={item._id} style={styles.card}>
    <Card
      style={styles.cardChild}
      onPress={() => navigation.navigate('PostView', {
        post: item,
      })}
    >
      <Card.Section
        bg-white
        content={[
          { text: item.content.length > 50 ? `${item.content.substring(0, 50)}...` : item.content, text70: true, grey10: true },
          { text: item?.authorId?.email, text90: true, grey40: true },
          { text: moment(item.createdAt).format('MMM Do YY'), text90: true, grey50: true },
        ]}
        style={{ padding: 20 }}
      />
      <View style={[HelperStyles.row,
        HelperStyles.paddingHorizontalBig,
        HelperStyles.bgWhite,
        HelperStyles.paddingBottomMed,
        HelperStyles.justifySpaceBetween,
      ]}
      >
        <View style={[HelperStyles.row,
          HelperStyles['w-50'],
        ]}
        >
          <View style={HelperStyles.paddingRightMed}>
            <Button
              text90
              link
              label={`Likes ${item.likes.length}`}
            />
          </View>

          <View>
            <Text
              text90
              grey40
              link
            >
              {`Comments ${item.comments.length}`}
            </Text>
          </View>
        </View>

        <View style={[HelperStyles['w-50'], HelperStyles.row, HelperStyles.justifyFlexEnd]}>
          {/* <Button text90 link label="Share" /> */}
        </View>
      </View>
    </Card>
  </View>
);

const arePropsEqual = (prevProps: IRenderPostProps, nextProps: IRenderPostProps) => (
  prevProps.item._id === nextProps.item._id
  && prevProps.item.content === nextProps.item.content
  && prevProps.item.comments.length === nextProps.item.comments.length
  && prevProps.item.likes.length === nextProps.item.likes.length
);

const MemoizedRenderPost = React.memo(RenderPost, arePropsEqual);

const FlatPostsList: FunctionComponent<IProps> = ({
  posts,
  navigation,
  renderHeader,
  renderFooter,
  loaded,
  handleGetPosts,
}: IProps) => (
  <FlatList
    data={posts}
    onRefresh={() => handleGetPosts(false)}
    refreshing={!loaded}
    style={HelperStyles.paddingHorizontalMed}
    renderItem={({ item }) => <MemoizedRenderPost item={item} navigation={navigation} />}
    ListHeaderComponent={renderHeader()}
    ListFooterComponent={renderFooter()}
    keyExtractor={(item) => item._id}
    ListEmptyComponent={loaded ? <Text style={{ textAlign: 'center' }}>No records found</Text> : null}
  />
);

export default FlatPostsList;
