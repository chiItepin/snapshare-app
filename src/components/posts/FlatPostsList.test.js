import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import FlatPostsList from './FlatPostsList';

jest.mock('react-redux', () => ({
  useSelector: () => ({
    _id: 'xxx',
  }),
}));

test('FlatPostsList renders and handles events', () => {
  const mockFn = jest.fn();
  const props = {
    navigation: {
      navigate: () => mockFn,
    },
    posts: [{
      _id: 'xxx',
      likes: [],
      comments: [],
      content: 'lorem ipsum',
      authorId: {
        email: 'xxx@xxx.com',
        image: '',
      },
      createdAt: 123,
    }],
    renderHeader: () => null,
    renderFooter: () => null,
    loaded: true,
    handleGetPosts: mockFn,
    handlePostLike: mockFn,
    handlePostShare: mockFn,
  };

  const component = render(
    <FlatPostsList {...props} />,
  );

  const [firsPostLikesBtn] = component.getAllByText(/Likes/);
  fireEvent.press(firsPostLikesBtn);

  const [firsPostShareBtn] = component.getAllByText(/Share/);
  fireEvent.press(firsPostShareBtn);

  expect(mockFn).toHaveBeenCalledTimes(2);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
