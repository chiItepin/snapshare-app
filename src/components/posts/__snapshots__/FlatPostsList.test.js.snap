// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatPostsList renders and handles events 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <Text
      grey40={true}
      style={
        Object {
          "textAlign": "center",
        }
      }
      text90={true}
    >
      No records found
    </Text>
  }
  ListFooterComponent={null}
  ListHeaderComponent={null}
  data={
    Array [
      Object {
        "_id": "xxx",
        "authorId": Object {
          "email": "xxx@xxx.com",
          "image": "",
        },
        "comments": Array [],
        "content": "lorem ipsum",
        "createdAt": "2021-09-13T18:00:00.440+00:00",
        "images": Array [],
        "likes": Array [],
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "paddingHorizontal": 12,
    }
  }
  testID="flat-posts-list"
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "backgroundColor": "rgb(240, 241, 244)",
            "borderRadius": 25,
            "marginBottom": 15,
            "marginTop": 5,
            "paddingHorizontal": 15,
            "paddingVertical": 15,
            "width": "100%",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "selected": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 25,
              "elevation": 2,
              "height": undefined,
              "marginBottom": 5,
              "opacity": 1,
              "overflow": "hidden",
              "shadowColor": "#A6ACB1",
              "shadowOffset": Object {
                "height": 5,
                "width": 0,
              },
              "shadowOpacity": 0.25,
              "shadowRadius": 12,
              "width": undefined,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flexDirection": "row",
                "justifyContent": "flex-start",
                "paddingHorizontal": 15,
                "paddingTop": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "width": "10%",
                }
              }
            >
              <View
                accessibilityLabel="Avatar"
                accessibilityRole="button"
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "borderRadius": 15,
                    "height": 30,
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": 30,
                  }
                }
              >
                <View
                  style={
                    Array [
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      Object {},
                      Object {},
                      Object {},
                      Array [
                        Object {
                          "alignItems": "center",
                          "borderRadius": 15,
                          "bottom": 0,
                          "justifyContent": "center",
                          "left": 0,
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                        },
                        Object {
                          "backgroundColor": "#FFF9E4",
                        },
                        false,
                      ],
                    ]
                  }
                  testID="undefined.container"
                >
                  <Text
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                          "textAlign": "left",
                        },
                        undefined,
                        undefined,
                        undefined,
                        undefined,
                        Object {},
                        undefined,
                        undefined,
                        undefined,
                        Array [
                          Object {
                            "fontSize": 9.6,
                          },
                          Object {
                            "backgroundColor": "transparent",
                            "color": "#20303C",
                            "lineHeight": undefined,
                          },
                          Object {
                            "color": "#20303C",
                          },
                        ],
                      ]
                    }
                    testID="undefined.label"
                  >
                    X
                  </Text>
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "width": "90%",
                }
              }
            >
              <View
                accessibilityRole="button"
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "color": "grey",
                    "fontWeight": "bold",
                    "opacity": 1,
                    "paddingLeft": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  grey30={true}
                  link={true}
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "textAlign": "left",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontSize": 12,
                        "fontWeight": "700",
                        "lineHeight": 16,
                      },
                      Object {
                        "color": "#6E7881",
                      },
                      undefined,
                      undefined,
                      Object {},
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                    ]
                  }
                  text90={true}
                >
                  xxx@xxx.com
                </Text>
              </View>
            </View>
          </View>
          <View
            bg-white={true}
            content={
              Array [
                Object {
                  "grey10": true,
                  "text": "lorem ipsum",
                  "text70": true,
                },
                Object {
                  "grey50": true,
                  "text": "Nov 13th 22",
                  "text90": true,
                },
              ]
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#FFFFFF",
                },
                undefined,
                undefined,
                undefined,
                Object {},
                Object {},
                Object {},
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Object {
                    "borderBottomLeftRadius": undefined,
                    "borderBottomRightRadius": undefined,
                    "borderTopLeftRadius": undefined,
                    "borderTopRightRadius": undefined,
                  },
                  Object {
                    "padding": 20,
                  },
                ],
              ]
            }
          >
            <View
              style={
                Array [
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Object {},
                  Object {},
                  Object {},
                  Array [
                    undefined,
                  ],
                ]
              }
              testID="undefined.contentContainer"
            >
              <Text
                grey10={true}
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "textAlign": "left",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontSize": 16,
                      "fontWeight": "400",
                      "lineHeight": 24,
                    },
                    Object {
                      "color": "#20303C",
                    },
                    undefined,
                    undefined,
                    Object {},
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                  ]
                }
                testID="undefined.text.0"
                text70={true}
              >
                lorem ipsum
              </Text>
              <Text
                grey50={true}
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "textAlign": "left",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontSize": 12,
                      "fontWeight": "700",
                      "lineHeight": 16,
                    },
                    Object {
                      "color": "#D2D6D8",
                    },
                    undefined,
                    undefined,
                    Object {},
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                  ]
                }
                testID="undefined.text.1"
                text90={true}
              >
                Nov 13th 22
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                },
                Object {
                  "paddingHorizontal": 24,
                },
                Object {
                  "backgroundColor": "white",
                },
                Object {
                  "paddingBottom": 12,
                },
                Object {
                  "justifyContent": "space-between",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                  },
                  Object {
                    "width": "100%",
                  },
                ]
              }
            >
              <View>
                <View
                  accessibilityRole="button"
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onLayout={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "borderRadius": 0,
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "minWidth": undefined,
                      "opacity": 1,
                      "paddingHorizontal": undefined,
                      "paddingRight": 12,
                      "paddingVertical": undefined,
                    }
                  }
                >
                  <Text
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                          "textAlign": "left",
                        },
                        undefined,
                        undefined,
                        undefined,
                        undefined,
                        Object {},
                        undefined,
                        undefined,
                        undefined,
                        Array [
                          Object {
                            "backgroundColor": "transparent",
                            "flex": 0,
                            "flexDirection": "row",
                            "fontFamily": "System",
                            "fontSize": 16,
                            "fontWeight": "400",
                            "lineHeight": 24,
                          },
                          Object {
                            "color": "#4D5963",
                          },
                          Object {},
                          Object {
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "400",
                            "lineHeight": 20,
                          },
                          undefined,
                        ],
                      ]
                    }
                    testID="undefined.label"
                  >
                    Likes 0
                  </Text>
                </View>
              </View>
              <View>
                <Text
                  grey20={true}
                  link={true}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "textAlign": "left",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "400",
                        "lineHeight": 20,
                      },
                      Object {
                        "color": "#4D5963",
                      },
                      undefined,
                      undefined,
                      Object {},
                      undefined,
                      undefined,
                      undefined,
                      Object {
                        "paddingRight": 12,
                      },
                    ]
                  }
                  text80={true}
                >
                  Comments 0
                </Text>
              </View>
              <View>
                <View
                  accessibilityRole="button"
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onLayout={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "borderRadius": 0,
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "minWidth": undefined,
                      "opacity": 1,
                      "paddingHorizontal": undefined,
                      "paddingRight": 12,
                      "paddingVertical": undefined,
                    }
                  }
                >
                  <Text
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                          "textAlign": "left",
                        },
                        undefined,
                        undefined,
                        undefined,
                        undefined,
                        Object {},
                        undefined,
                        undefined,
                        undefined,
                        Array [
                          Object {
                            "backgroundColor": "transparent",
                            "flex": 0,
                            "flexDirection": "row",
                            "fontFamily": "System",
                            "fontSize": 16,
                            "fontWeight": "400",
                            "lineHeight": 24,
                          },
                          Object {
                            "color": "#4D5963",
                          },
                          Object {},
                          Object {
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "400",
                            "lineHeight": 20,
                          },
                          undefined,
                        ],
                      ]
                    }
                    testID="undefined.label"
                  >
                    Share
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
